How do you update a custom user storage provider jar that includes a version number?

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

I have a User Storage Provider jar that I wrote to use our corporate sign in databse from keycloak which I wrote when using keycloak 19.0.2. I have since updated keycloak to 23.0.4 and had to update the user storage provider jar. When I did so, I updated the version number, which is also in the name of the jar.

However, I am now unable to use the new jar as keycloak's build process complains that the old jar is not found and crashes :(

I have had to symlink the new jar to the old name to get keycloak to build and start, how do I get around this limitation?

Note: I have been using the Quarkus version for 19.0.2 and, of course, for 23.0.4

### Version

23.0.4

### Expected behavior

I should be able to replace the old custom provider jar in the providers directory with the new version without having to symlink the new jar file to have the old name :(

### Actual behavior

keycloak will not build or start when you update a jar in providers (a user storage provider in this case) by deleting the old version and adding the new version. You have to symlink , or rename, the new jar to have the same name as the old version :(

### How to Reproduce?

1. Have a user storage provider with a version in the name, such as keycloak-user-store-1.0.0-SNAPSHOT.jar.
2. Build and start the keycloak server
3. Shut down the keycloak server.
4. Then attempt to update the user storage provider to a new version with a different version number, such as keycloak-user-store-2.0.0-SNAPSHOT.jar, and delete the old keycloak-user-store-1.0.0-SNAPSHOT.jar
5. Start the server

### Anything else?

_No response_

@shawkins The stacktrace is:
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.RuntimeException: Failed to open /local/app/kc/keycloak-23.0.6/lib/../providers/els-user-store-24.02.50-SNAPSHOT.jar
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:172)
        at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
        at io.quarkus.bootstrap.runner.JarResource.getResourceData(JarResource.java:72)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:99)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
        at org.keycloak.quarkus.runtime.Environment.isWindows(Environment.java:156)
        at org.keycloak.quarkus.runtime.Environment.getCommand(Environment.java:95)
        at org.keycloak.quarkus.runtime.cli.Picocli.createCommandLine(Picocli.java:448)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:97)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:107)
        ... 6 more
Caused by: java.nio.file.NoSuchFileException: /local/app/kc/keycloak-23.0.6/lib/../providers/els-user-store-24.02.50-SNAPSHOT.jar
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148)
        at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1428)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:718)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:252)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:181)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:346)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:33)
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
        ... 15 more

@shawkins , I removed commons lang3 from the jar (build.gradle), rebuilt it and tried again. It still crashes :(
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.RuntimeException: Failed to open /local/app/kc/keycloak-23.0.6/lib/../providers/els-user-store-24.02.50-SNAPSHOT.jar
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:172)
        at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
        at io.quarkus.bootstrap.runner.JarResource.getResourceData(JarResource.java:72)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:99)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
        at org.keycloak.quarkus.runtime.Environment.isWindows(Environment.java:156)
        at org.keycloak.quarkus.runtime.Environment.getCommand(Environment.java:95)
        at org.keycloak.quarkus.runtime.cli.Picocli.createCommandLine(Picocli.java:448)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:97)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:107)
        ... 6 more
Caused by: java.nio.file.NoSuchFileException: /local/app/kc/keycloak-23.0.6/lib/../providers/els-user-store-24.02.50-SNAPSHOT.jar
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148)
        at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1428)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:718)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:252)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:181)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:346)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:33)
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
        ... 15 more

--- a/docs/documentation/server_development/topics/providers.adoc
+++ b/docs/documentation/server_development/topics/providers.adoc
@@ -223,9 +223,11 @@ If your provider needs additional dependencies not already provided by Keycloak 
 [NOTE]
 ====
 Provider JARs are not loaded in isolated classloaders, so do not include resources or classes in your provider JARs that conflict with built-in resources or classes. 
-In particular the inclusion of an application.properties file will cause auto-build to fail if the provider JAR is removed. 
-If you have included conflicting classes, you will see a split package warning in the start log for the server. That should be resolved by removing or repackaging the offending classes. 
-However there is no warning if you have conflicting resource files. You should either ensure that your JAR's resource files have path names that contain something unique to that provider, 
+In particular the inclusion of an application.properties file or overriding the commons-lang3 dependency will cause auto-build to fail if the provider JAR is removed. 
+If you have included conflicting classes, you may see a split package warning in the start log for the server. Unfortunately not all built-in lib jars are checked by the split package warning logic,
+so you'll need to check the lib directory JARs before bundling or including a transitive dependency. Should there be a conflict, that can be resolved by removing or repackaging the offending classes.
+
+There is no warning if you have conflicting resource files. You should either ensure that your JAR's resource files have path names that contain something unique to that provider, 
 or you can check for the existence of `some.file` in the JAR contents under the `"install root"/lib/lib/main` directory with something like:
  
 [source,bash]

